File: wear/app/src/main/java/app/imuatrak/wear/models/Session.kt
package app.imuatrak.wear.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

// ── Session model — mirrors src/models/index.ts exactly ──────────────────────
// Field names must match the TypeScript JSON schema so the phone app can
// deserialise transferred files without any conversion.

@Serializable
data class WatchSession(
    val id: String,
    val userId: String = "",
    val schemaVersion: Int = 1,
    val source: String = "android-wear",
    val appVersion: String = "0.1.0",
    val craftType: String,
    val startedAt: String,
    val endedAt: String,
    val totals: WatchTotals,
    val hr: WatchHrSummary,
    val splits: List<WatchSplit>,
    val sideSwitches: List<WatchSideSwitch> = emptyList(),
    val weather: WatchWeatherSummary? = null,
    val trackSummary: List<WatchTrackPoint>,
    val isPublic: Boolean? = false,
)

@Serializable
data class WatchTotals(
    val distanceMeters: Double,
    val durationSec: Double,
    val movingDurationSec: Double,
    val avgPaceSecPerKm: Double,
    val avgSpeedMps: Double,
    val maxSpeedMps: Double,
    val strokeCount: Int,
    val avgStrokeRate: Double,
    val calories: Double = 0.0,
    val elevationGainM: Double,
) {
    companion object {
        fun empty() = WatchTotals(0.0,0.0,0.0,0.0,0.0,0.0,0,0.0,0.0,0.0)
    }
}

@Serializable
data class WatchHrSummary(
    val avg: Int,
    val max: Int,
    val zones: List<WatchHrZone>,
) {
    companion object {
        fun empty(): WatchHrSummary {
            val bounds = listOf(0 to 120, 120 to 140, 140 to 160, 160 to 175, 175 to 220)
            return WatchHrSummary(0, 0, bounds.mapIndexed { i, (mn, mx) ->
                WatchHrZone(i, mn, mx, 0.0)
            })
        }
    }
}

@Serializable
data class WatchHrZone(val zone: Int, val minBpm: Int, val maxBpm: Int, var timeSec: Double)

@Serializable
data class WatchSplit(
    val index: Int,
    val distanceM: Double,
    val durationSec: Double,
    val avgHr: Double,
    val avgStrokeRate: Double,
    val avgSpeedMps: Double,
)

@Serializable
data class WatchSideSwitch(
    val tSec: Double,
    val detectedSide: String,
    val confidence: Double,
    val source: String,
)

@Serializable
data class WatchTrackPoint(
    val t: Double,
    val lat: Double,
    val lon: Double,
    val altM: Double,
    val speedMps: Double,
    val hr: Int? = null,
    val strokeRate: Double? = null,
    val cadenceConfidence: Double? = null,
)

@Serializable
data class WatchWeatherSummary(val start: WatchWeatherSample, val samples: List<WatchWeatherSample>)

@Serializable
data class WatchWeatherSample(
    val tSec: Double,
    val windMps: Double,
    val windDeg: Double,
    val gustMps: Double,
    val airTempC: Double,
    val pressureHpa: Double,
    val conditions: String? = null,
)

File: targets/watch/Models/Session.swift
import Foundation

// ── Session model — mirrors src/models/index.ts exactly ──────────────────────
// Field names and JSON keys must stay in sync with the TypeScript schema so
// that the phone app can deserialise files transferred from the watch.

struct WatchSession: Codable {
    let id: String
    let userId: String
    let schemaVersion: Int
    let source: String          // "ios-watch"
    let appVersion: String
    let craftType: String
    let startedAt: String       // ISO-8601
    let endedAt: String
    var totals: WatchTotals
    var hr: WatchHrSummary
    var splits: [WatchSplit]
    let sideSwitches: [WatchSideSwitch]
    var weather: WatchWeatherSummary?
    var trackSummary: [WatchTrackPoint]
    var isPublic: Bool?
}

struct WatchTotals: Codable {
    var distanceMeters: Double
    var durationSec: Double
    var movingDurationSec: Double
    var avgPaceSecPerKm: Double
    var avgSpeedMps: Double
    var maxSpeedMps: Double
    var strokeCount: Int
    var avgStrokeRate: Double
    var calories: Double
    var elevationGainM: Double
}

struct WatchHrSummary: Codable {
    var avg: Int
    var max: Int
    var zones: [WatchHrZone]
}

struct WatchHrZone: Codable {
    let zone: Int
    let minBpm: Int
    let maxBpm: Int
    var timeSec: Double
}

struct WatchSplit: Codable {
    let index: Int
    let distanceM: Double
    let durationSec: Double
    var avgHr: Double
    var avgStrokeRate: Double
    var avgSpeedMps: Double
}

struct WatchSideSwitch: Codable {
    let tSec: Double
    let detectedSide: String
    let confidence: Double
    let source: String
}

struct WatchTrackPoint: Codable {
    let t: Double
    let lat: Double
    let lon: Double
    let altM: Double
    let speedMps: Double
    var hr: Int?
    var strokeRate: Double?
    var cadenceConfidence: Double?
}

struct WatchWeatherSummary: Codable {
    let start: WatchWeatherSample
    let samples: [WatchWeatherSample]
}

struct WatchWeatherSample: Codable {
    let tSec: Double
    let windMps: Double
    let windDeg: Double
    let gustMps: Double
    let airTempC: Double
    let pressureHpa: Double
    let conditions: String?
}

// ── Helpers ───────────────────────────────────────────────────────────────────

extension WatchTotals {
    static func empty() -> WatchTotals {
        WatchTotals(distanceMeters: 0, durationSec: 0, movingDurationSec: 0,
                    avgPaceSecPerKm: 0, avgSpeedMps: 0, maxSpeedMps: 0,
                    strokeCount: 0, avgStrokeRate: 0, calories: 0, elevationGainM: 0)
    }
}

extension WatchHrSummary {
    static func empty() -> WatchHrSummary {
        let bounds = [(0,120),(120,140),(140,160),(160,175),(175,220)]
        let zones = bounds.enumerated().map { i, b in
            WatchHrZone(zone: i, minBpm: b.0, maxBpm: b.1, timeSec: 0)
        }
        return WatchHrSummary(avg: 0, max: 0, zones: zones)
    }
}
